serde/src/ser/mod.rs

```diff
@@ -202,53 +212,60 @@ pub trait Serializer {
     ///
     /// Callees of this method need to construct a `SeqVisitor`, which iterates through each item
     /// in the sequence.
-    fn serialize_seq<V>(&mut self, visitor: V) -> Result<(), Self::Error>
-        where V: SeqVisitor;
+    fn serialize_seq(&mut self, len: Option<usize>) -> Result<Self::SeqState, Self::Error>;
 
     /// Serializes a sequence element.
     fn serialize_seq_elt<T>(&mut self, value: T) -> Result<(), Self::Error>
         where T: Serialize;
 
+    /// Finish serializing a sequence.
+    fn serialize_seq_end(&mut self, len: Option<usize>, state: Self::SeqState) -> Result<(), Self::Error>;
+
     /// Serializes a tuple.
     ///
     /// By default this serializes a tuple as a sequence.
     #[inline]
-    fn serialize_tuple<V>(&mut self, visitor: V) -> Result<(), Self::Error>
-        where V: SeqVisitor,
+    fn serialize_tuple(&mut self, len: usize) -> Result<Self::SeqState, Self::Error>
     {
-        self.serialize_seq(visitor)
+        self.serialize_seq(Some(len))
     }
 
     /// Serializes a tuple element.
     ///
     /// By default, tuples are serialized as a sequence.
     #[inline]
     fn serialize_tuple_elt<T>(&mut self, value: T) -> Result<(), Self::Error>
-        where T: Serialize
-    {
+        where T: Serialize {
         self.serialize_seq_elt(value)
     }
 
+    /// Finishes serialization of a tuple.
+    ///
+    /// By default, tuples are serialized as a sequence.
+    #[inline]
+    fn serialize_tuple_end(&mut self, len: usize, state: Self::SeqState) -> Result<(), Self::Error> {
+        self.serialize_seq_end(Some(len), state)
+    }
+
     /// Serializes a fixed-size array.
     ///
     /// By default this serializes an array as a sequence.
     #[inline]
-    fn serialize_fixed_size_array<V>(&mut self, visitor: V) -> Result<(), Self::Error>
-        where V: SeqVisitor,
+    fn serialize_fixed_size_array(&mut self, size: usize) -> Result<Self::SeqState, Self::Error>
```

Comment:
Let's remove this method. [No Serializer on GitHub uses it.](https://github.com/search?l=rust&q=serialize_fixed_size_array&type=Code&utf8=%E2%9C%93) Why would anything treat it differently from `serialize_seq(Some(size))`?